src/api/app/models/event/request.rb

```diff
@@ -77,12 +77,36 @@ def source_maintainers
       action_maintainers('sourceproject', 'sourcepackage')
     end
 
-    def target_watchers
-      find_watchers('targetproject')
+    def source_watchers
+      # TODO: remove old `find_watchers` implementation after rolling out `new_watchlist` from beta program
+      find_watchers('sourceproject') +
+        payload['actions'].map { |action| action['sourceproject'] }
+                          .map { |source_project_name| Project.find_by_name(source_project_name) }
+                          .compact.map(&:watched_items)
+                          .flatten.map(&:user)
     end
 
-    def source_watchers
-      find_watchers('sourceproject')
+    def target_watchers
+      # TODO: remove old `find_watchers` implementation after rolling out `new_watchlist` from beta program
+      find_watchers('targetproject') +
+        payload['actions'].map { |action| action['targetproject'] }
+                          .map { |target_project_name| Project.find_by_name(target_project_name) }
+                          .compact.map(&:watched_items)
+                          .flatten.map(&:user)
+    end
+
+    def source_package_watchers
+      payload['actions'].map { |action| action['sourcepackage'] }
+                        .map { |source_package_name| Package.find_by_name(source_package_name) }
```

Comment:
This won't find a specific package with the provided package name. It needs to also be scoped to the project.